com.unity.netcode.gameobjects/Runtime/Messaging/DeferredMessageManager.cs

```diff
@@ -0,0 +1,150 @@
+using System.Collections.Generic;
+using Unity.Collections;
+using Time = UnityEngine.Time;
+
+namespace Unity.Netcode
+{
+    internal class DeferredMessageManager : IDeferredMessageManager
+    {
+        protected struct TriggerData
+        {
+            public FastBufferReader Reader;
+            public MessageHeader Header;
+            public ulong SenderId;
+            public float Timestamp;
+            public int SerializedHeaderSize;
+        }
+        protected struct TriggerInfo
+        {
+            public float Expiry;
+            public NativeList<TriggerData> TriggerData;
+        }
+
+        protected readonly Dictionary<IDeferredMessageManager.TriggerType, Dictionary<ulong, TriggerInfo>> m_Triggers = new Dictionary<IDeferredMessageManager.TriggerType, Dictionary<ulong, TriggerInfo>>();
+
+        private readonly NetworkManager m_NetworkManager;
+
+        internal DeferredMessageManager(NetworkManager networkManager)
+        {
+            m_NetworkManager = networkManager;
+        }
+
+        /// <summary>
+        /// Defers processing of a message until the moment a specific networkObjectId is spawned.
+        /// This is to handle situations where an RPC or other object-specific message arrives before the spawn does,
+        /// either due to it being requested in OnNetworkSpawn before the spawn call has been executed
+        ///
+        /// There is a one second maximum lifetime of triggers to avoid memory leaks. After one second has passed
+        /// without the requested object ID being spawned, the triggers for it are automatically deleted.
+        /// </summary>
+        public virtual unsafe void DeferMessage(IDeferredMessageManager.TriggerType trigger, ulong key, FastBufferReader reader, ref NetworkContext context)
+        {
+            if (!m_Triggers.ContainsKey(trigger))
+            {
+                m_Triggers[trigger] = new Dictionary<ulong, TriggerInfo>();
+            }
+
+            var triggers = m_Triggers[trigger];
+            if (!triggers.ContainsKey(key))
+            {
+                triggers[key] = new TriggerInfo
+                {
+                    Expiry = Time.realtimeSinceStartup + m_NetworkManager.NetworkConfig.SpawnTimeout,
+                    TriggerData = new NativeList<TriggerData>(Allocator.Persistent)
+                };
+            }
+
+            triggers[key].TriggerData.Add(new TriggerData
+            {
+                Reader = new FastBufferReader(reader.GetUnsafePtr(), Allocator.Persistent, reader.Length),
```

Comment:
I am not sure about this, but doesn't this allocate each time you are "deferring" a message?

Reply:
Not in the way you're thinking. FastBufferReader is a struct so it's allocated on the stack. It contains only an unsafe pointer so its size is only one word. Its underlying data is allocated using Unity's native allocators (same ones used for NativeList and NativeArray) - so yes, there's an allocation, but it's a low-level malloc and doesn't create garbage.

It's unavoidable though because the existing reader was allocated with `Allocator.TempJob` and those have a maximum lifetime of 4 frames. Since the deferred messages can last up to 60 frames (or more if `NetworkConfig.SpawnTimeout` is increased) it has to be moved to `Allocator.Persistent`. The whole buffer copy isn't ideal... But this is here to handle an edge-case scenario so it's not like it'll be happening constantly.